### PR TITLE
Weak interior handles

### PR DESCRIPTION
**Pull request by davidwrighton**
Add a new type of handle (HNDTYPE_WEAK_INTERIOR_POINTER) and a use of it to make storage of managed class objects consistent between collectible and non-collectible variants

### CODE REVIEW

File: src/coreclr/vm/loaderallocator.cpp [hunk 1]
```diff
@@ -2198,3 +2217,53 @@ PTR_OnStackReplacementManager LoaderAllocator::GetOnStackReplacementManager()
...
+    {
+        GCX_FORBID();
+        // Allocating a weak interior handle is a tricky thing.
+        // 1. If there are multiple weak interior handles that point at a given interior pointer location, there will be heap corruption
+        // 2. If the interior handle is created, but not registered for cleanup, it is a memory leak
+        // 3. Since the weak interior handle doesn't actually keep the object alive, it needs to be kept alive by some other means
+        //
+        // We work around these details by the following means
+        // 1. We use a LOADERHANDLE to keep the object alive until the LoaderAllocator is freed.
+        // 2. We hold the crstLoaderAllocatorLock, and double check to wnsure that the data is ready to be filled in
```
Comment (teo-tsirpanis) [inline src/coreclr/vm/loaderallocator.cpp:2251]:
```suggestion
        // 2. We hold the crstLoaderAllocatorLock, and double check to ensure that the data is ready to be filled in
```

**Comment by Maoni0** [comment]
I presume the extra info for these handles must always point to the middle of the primary target of this handle (it's not obvious from the caller I'm seeing in loaderallocator.cpp). we should add some verification code that this is the case.

I thought I'd mention we do have a `GCHeap::GetContainingObject" that gets you from an interior pointer to its containing object, in case this is useful to you. 

otherwise the change looks ok to me.

**Comment by davidwrighton** [comment]
> I presume the extra info for these handles must always point to the middle of the primary target of this handle (it's not obvious from the caller I'm seeing in loaderallocator.cpp). we should add some verification code that this is the case.
Ok, done.

> 
> I thought I'd mention we do have a `GCHeap::GetContainingObject" that gets you from an interior pointer to its containing object, in case this is useful to you.
Unfortunately my understanding is that api can only be called when the runtime is suspended due to race conditions, so I can't use this in any of the cases I'm working with.

File: src/coreclr/gc/handletablescan.cpp [hunk 2]
```diff
@@ -992,6 +1001,23 @@ void BlockVerifyAgeMapForBlocksWorker(uint32_t *pdwGen, uint32_t dwClumpMask, Sc
...
                         }
                     }
+                    if (uType == HNDTYPE_WEAK_INTERIOR_POINTER)
+                    {
+                        PTR_uintptr_t pUserData = HandleQuickFetchUserDataPointer((OBJECTHANDLE)pValue);
+
+                        // if we did then copy the value
+                        if (pUserData)
+                        {
+                            TADDR pObjectInteriorPointer = **dac_cast<DPTR(DPTR(TADDR))>(pUserData);
```
Comment (jkotas) [inline src/coreclr/gc/handletablescan.cpp:1011]: Nit: None of the surrounding code is DACized so dac_cast is unnecessary complication.

Comment (davidwrighton) [reply src/coreclr/gc/handletablescan.cpp:1011]: Yeah, I thought that VerifyHeap worked in the DAC, but it turns out that it is an entirely parallel managed implementation these days. I'll pull that out if you'd like.

Comment (jkotas) [reply src/coreclr/gc/handletablescan.cpp:1011]: There is a build break introduced by the `dac_cast` macro. My guess is that the easiest way to fix it is to delete the use here.

Comment (davidwrighton) [reply src/coreclr/gc/handletablescan.cpp]: Done

File: src/coreclr/vm/loaderallocator.cpp [hunk 3]
```diff
@@ -2198,3 +2217,56 @@ PTR_OnStackReplacementManager LoaderAllocator::GetOnStackReplacementManager()
...
+        if (*pField == NULL)
+        {
+            // Allocating a weak interior handle is a tricky thing.
+            // 1. If there are multiple weak interior handles that point at a given interior pointer location, there will be heap corruption
+            // 2. If the interior handle is created, but not registered for cleanup, it is a memory leak
+            // 3. Since the weak interior handle doesn't actually keep the object alive, it needs to be kept alive by some other means
+            //
+            // We work around these details by the following means
+            // 1. We use a LOADERHANDLE to keep the object alive until the LoaderAllocator is freed.
+            // 2. We hold the crstLoaderAllocatorLock, and double check to ensure that the data is ready to be filled in
```
Comment (jkotas) [inline src/coreclr/vm/loaderallocator.cpp:2253]:
```suggestion
            // 2. We hold the m_crstLoaderAllocator lock, and double check to ensure that the data is ready to be filled in
```
Nit

File: src/coreclr/vm/typedesc.h [hunk 4]
```diff
@@ -163,6 +163,25 @@ class TypeDesc
...
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+            MODE_COOPERATIVE;
+        }
+        CONTRACTL_END;
+
+        const RUNTIMETYPEHANDLE handle = m_hExposedClassObject;
+        OBJECTREF retVal = ObjectToOBJECTREF((Object*)handle);
```
Comment (jkotas) [inline src/coreclr/vm/typedesc.h:180]:
```suggestion
        OBJECTREF retVal = ObjectToOBJECTREF(handle);
```
RUNTIMETYPEHANDLE can be defined as `Object*` to save casts like this. It is actually Object* now.

**Review by jkotas — APPROVED** [review]
There is a potential follow up codegen improvement for collectible assemblies:

We can change `getRuntimeTypePointer` to return `InfoAccessType` in addition to the pointer itself and allow JIT to fetch the Type object by a simple dereference instead of calling the helper in collectible code.

cc @EgorBo

**Pull request merged**